src/openrct2/world/Banner.cpp

```diff
@@ -253,57 +254,50 @@ void fix_duplicated_banners()
 {
     // For each banner in the map, check if the banner index is in use already, and if so, create a new entry for it
     std::vector<bool> activeBanners;
+    activeBanners.resize(MAX_BANNERS);
+
     for (int y = 0; y < MAXIMUM_MAP_SIZE_TECHNICAL; y++)
     {
         for (int x = 0; x < MAXIMUM_MAP_SIZE_TECHNICAL; x++)
         {
-            auto tileElement = map_get_first_element_at(TileCoordsXY{ x, y }.ToCoordsXY());
-            if (tileElement != nullptr)
+            const auto bannerPos = TileCoordsXY{ x, y }.ToCoordsXY();
+            for (auto* bannerElement : OpenRCT2::TileElementsView<BannerElement>(bannerPos))
             {
-                do
+                auto bannerIndex = bannerElement->GetIndex();
+                if (bannerIndex == BANNER_INDEX_NULL)
+                    continue;
+
+                if (activeBanners[bannerIndex])
                 {
-                    // TODO: Handle walls and large-scenery that use banner indices too. Large scenery can be tricky, as they
-                    // occupy multiple tiles that should both refer to the same banner index.
-                    if (tileElement->GetType() == TILE_ELEMENT_TYPE_BANNER)
+                    log_info(
+                        "Duplicated banner with index %d found at x = %d, y = %d and z = %d.", bannerIndex, x, y,
+                        bannerElement->base_height);
+
+                    // Banner index is already in use by another banner, so duplicate it
+                    auto newBanner = CreateBanner();
+                    if (newBanner == nullptr)
+                    {
+                        log_error("Failed to create new banner.");
+                        continue;
```

Comment:
I think if it comes down to this it should probably just return, any further attempt will likely return nullptr.